### PR TITLE
multiple CIDR and configurable hostsubnetlength support 

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
First draft of adding multiple CIDR and configurable hostsubnetlengh in ovn-kubernetes networkwork plugin.

multiple cidrs are passed on the command line with  --cluster-subnet flag in the form

`--cluster-subnet [cidr]-[hostsubnetlength],...`
for example I used the below for testing
 
`-cluster-subnet "10.132.0.0/26-4,10.133.0.0/26-4"`

the hostsubnetlength refers to the number of ip addresses each node has for the allocation of pods in the above the nodes allocated in the above ranges would have /28 addresses to allocate pods in (I made it small for testing purposes) without this PR the hostsubnetlength was hard-coded to 8 so nodes had /24

### CODE REVIEW

**Comment by shettyg** [comment]
Please add a Signed-off-by and limit the first line of your commit message to something smaller (just a subject).

What does `-cluster-subnet "10.132.0.0/26-4,10.133.0.0/26-4"` mean? Specifically, "-4". Can you please elaborate? (I haven't looked at code)

**Comment by dcbw** [comment]
@shettyg I believe the "-4" is how we're specifying HostSubnetLength, which is currently hardcoded.  When we add multiple CIDRs each of the blocks might have a different HostSubnetLength depending on how large the space is that the admin can carve out.

**Comment by dcbw** [comment]
@JacobTanenbaum also, a couple bugs showed up in the Travis golint tests.  Looks like you may have missed an argument change for initGateways() in the Windows code and then I think you want to add codedoc and maybe change the name of ClusterIPEntry.

**Comment by shettyg** [comment]
>@shettyg I believe the "-4" is how we're specifying HostSubnetLength, which is currently hardcoded. When we add multiple CIDRs each of the blocks might have a different HostSubnetLength depending on how large the space is that the admin can carve out.

I guess, I still don't understand. Lets take the example of 10.132.0.0/26-4.  So If we just take "10.132.0.0/26", it tells me the network is 26 bits and 6 bits is reserved for host. So that gives me (2^6 - 2) IP addresses in that subnet (-2 is because one is broadcast address and .0 is not a valid IP address).  Now why do I need to, in addition, say -4? 

And what do you mean by ""-4" is how we're specifying HostSubnetLength, which is currently hardcoded"? Where do we hardcode it?

**Comment by JacobTanenbaum** [comment]
@shettyg in the current code the bits reserved for the host is 8. 
It gets set in  go-controller/cmd/ovnkube/ovnkube.go

`clusterController.HostSubnetLength = 8`

We might want to change it to any number based on the sizes of the network space or as a way to limit the number of objects deployed to any specific node

**Comment by dcbw** [comment]
> I guess, I still don't understand. Lets take the example of 10.132.0.0/26-4. So If we just take "10.132.0.0/26", it tells me the network is 26 bits and 6 bits is reserved for host. So that gives me (2^6 - 2) IP addresses in that subnet (-2 is because one is broadcast address and .0 is not a valid IP address). Now why do I need to, in addition, say -4?

@shettyg we need two things to assign pod subnets to nodes:

1) the CIDR space from which hosts will be allocated their slice of the CIDR (ClusterCIDR, eg 10.132.0.0/26)
2) how many addresses each host should get (HostSubnetLength, eg 8)

The problem we have is that not all setups can dedicate the whole CIDR to the cluster; say they have legacy resources in part of 10./8 or 172.16/16 or whatever that they cannot move.  So we need to punch holes into the overall ClusterCIDR for this, and it turns out to be easier to just give a list of allowed CIDRs + HostSubnetLengths rather than blacklisting or whatever.

Does that make more sense?  Hopefully I've cleared up some of the confusion.

**Comment by danwinship** [comment]
> I guess, I still don't understand. Lets take the example of 10.132.0.0/26-4. So If we just take "10.132.0.0/26", it tells me the network is 26 bits and 6 bits is reserved for host.

No, it's saying that 10.132.0.0/26 is a range that's available to the whole cluster, and each node that gets allocated out of this space will get a (32-4=) /28 within it. So it provides for 4 nodes, with the ranges 10.132.0.0/28, 10.132.0.16/28, 10.132.0.32/28, and 10.132.0.48/28.

**Comment by danwinship** [comment]
The cidr+hostSubnetLength notation is based on how OpenShift SDN does it, but I have always felt that the way we do it is backwards, and it would have made more sense to specify the prefix length of the subnet that each host will get rather than the number of bits reserved for the host. So eg, `10.132.0.0/26-28` rather than `10.132.0.0/26-4` here. Not sure the hyphen is the best punctuation there either since it makes it look like it's a range, which it isn't. Maybe `10.132.0.0/26/28` ?

**Comment by shettyg** [comment]
>Does that make more sense? Hopefully I've cleared up some of the confusion.

Yeah. I understand it now.

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 1]
```diff
@@ -279,3 +281,29 @@ func runOvnKube(ctx *cli.Context) error {
 
 	return nil
 }
+
+func parseClusterSubnetEntries(clusterSubnetCmd string) ([]ovncluster.CIDRNetworkEntry, error) {
```
Comment (squeed) [inline go-controller/cmd/ovnkube/ovnkube.go]: can you add a quick docstring? (and a unit test :-) )

**Comment by JacobTanenbaum** [comment]
@danwinship changed the way CIDR and hostsubnetlength are defined 
`--cluster-subnet "10.132.0.0/26/28,10.133.0.0/26/28"`

@squeed added unit testing

**Comment by shettyg** [comment]
>--cluster-subnet "10.132.0.0/26/28,10.133.0.0/26/28"

We still want it to work if someone does not specify the extra argument - otherwise it is a regression. A quick glance of code suggested it is not the case?

File: go-controller/pkg/cluster/master.go [hunk 2]
```diff
@@ -344,18 +353,25 @@ func (cluster *OvnClusterController) addNode(node *kapi.Node) error {
...
+			if err != nil {
+				_ = possibleSubnet.ReleaseNetwork(sn)
+				return fmt.Errorf("Error creating subnet %s for node %s: %v", sn.String(), node.Name, err)
+			}
+			logrus.Infof("Created HostSubnet %s", sn.String())
+			return nil
+		}
 	}
-	logrus.Infof("Created HostSubnet %s", sn.String())
 	return nil
```
Comment (dcbw) [inline go-controller/pkg/cluster/master.go]: @JacobTanenbaum shouldn't we return an error here saying that all subnets have been exhausted?

File: go-controller/cmd/ovnkube/ovnkube.go [hunk 3]
```diff
@@ -279,3 +281,41 @@ func runOvnKube(ctx *cli.Context) error {
 
 	return nil
 }
+
+// parseClusterSubnetEntries returns the parsed set of CIDRNetworkEntries passed by the user on the command line
+// These entries define the clusters network space by specifying a set of CIDR and netmaskas the SDN can allocate
+// addresses from.
+func parseClusterSubnetEntries(clusterSubnetCmd string) ([]ovncluster.CIDRNetworkEntry, error) {
```
Comment (squeed) [inline go-controller/cmd/ovnkube/ovnkube.go]: Oh, I just remembered that, in the code for the CNI host-local allocator, we detect if the administrator accidentally configures overlapping subnets.

There's a simple implementation here: https://github.com/containernetworking/plugins/blob/master/plugins/ipam/host-local/backend/allocator/range.go#L131

Do you want to add this? It could be a separate PR, if you like.

Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: @squeed I'm fine with it being a follow-up PR.

Comment (dcbw) [reply go-controller/cmd/ovnkube/ovnkube.go]: @squeed Jacob added code and testcases to do this.

**Comment by dcbw** [comment]
> We still want it to work if someone does not specify the extra argument - otherwise it is a regression. A quick glance of code suggested it is not the case?

@shettyg Jacob added code and testcases to ensure that the old format still works and defaults to 8-bits of HostSubnetLength.  Look better now?

File: go-controller/cmd/ovn-kube-util/app/gateway_init.go [hunk 4]
```diff
@@ -92,7 +92,7 @@ func initGateway(context *cli.Context) error {
 		return fmt.Errorf("One of physical-interface or bridge-interface has to be specified")
 	}
 
-	return util.GatewayInit(clusterIPSubnet, nodeName, physicalIP,
+	return util.GatewayInit([]string{clusterIPSubnet}, nodeName, physicalIP,
```
Comment (squeed) [inline go-controller/cmd/ovn-kube-util/app/gateway_init.go]: hmm - I'm confused - why is this hardcoed to a list with one entry?

**Comment by squeed** [comment]
One question, which is probably me being confused, otherwise LGTM.

File: go-controller/cmd/ovn-kube-util/app/gateway_init.go (hunk 4 above)
Comment (JacobTanenbaum) [reply go-controller/cmd/ovn-kube-util/app/gateway_init.go]: @squeed this util function is called with an array in other places but in this context there is only a single element. Should I make the ovn-kube-utils support multiple subnets as well?

Comment (squeed) [reply go-controller/cmd/ovn-kube-util/app/gateway_init.go]: no need - I just wanted to understand. If we need it, we can add it later.

**Comment by shettyg** [comment]
We are missing documentation in docs/ovnkube.1 with new definition.  I will also run through the pull request through vagrant to make sure there are no regressions.

**Comment by rajatchopra** [comment]
/test integration

We should try to revive the integration tests as well meanwhile.

**Comment by dcbw** [comment]
@JacobTanenbaum can you update docs/ovnkube.1?

**Comment by shettyg** [comment]
I did not test new code. But I did not see any regressions. /LGTM

**Comment by dcbw** [comment]
@shettyg yeah, I tested this PR in Vagrant as well and was able to start an nginx pod without problems.

**Comment by dcbw** [comment]
@JacobTanenbaum thanks, /lgtm now
